Span-based RemoveDotSegments (#1448)

cesarbs opened the pull request:
#1448 

Working on perf, but already looks better than string-based implementation:

```
           Method |        Mean |    StdDev |        Op/s | Scaled | Scaled-StdDev | Allocated |
----------------- |------------ |---------- |------------ |------- |-------------- |---------- |
 StringNoSegments |  43.8869 ns | 0.2232 ns | 22785819.23 |   1.00 |          0.00 |       0 B |
   StringSegments | 168.3321 ns | 2.6535 ns |   5940636.5 |   3.84 |          0.06 |      56 B |
   SpanNoSegments |  38.3205 ns | 0.3611 ns | 26095668.18 |   0.87 |          0.01 |       0 B |
     SpanSegments |  65.1648 ns | 0.4828 ns | 15345719.56 |   1.48 |          0.01 |       0 B |
```

- [x] Will make changes to `Frame` after #1489.

pakrym commented on src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/PathNormalizer.cs:
```diff
@@ -1,12 +1,128 @@
+
+                while (src < end)
+                {
+                    if (end - src >= 3 && *src == '.' && *(src + 1) == '.' && *(src + 2) == '/')
```
What if we group checks by length and cache characters like:
```
if (end - src >= 3)
{
 var c1 = *src;
 var c2 = *(src + 1);
 var c3 = *(src + 2);
 if (c1 == '.' && c2 == '.' && c3 == '/') {}
 else if (c1 == ...) {}
}
```

pakrym replied:
Also we had chars cached as bytes in Frame would it help here if we compare with const bytes vs chars?

cesarbs replied:
Will try those. Also thinking of putting segments in `int`s for faster fetch and comparison.

cesarbs commented:
@pakrym Caching chars has a big benefit but code is a bit messier (needs some dupe checks and `continue`s everywhere - check it out).

```
           Method |        Mean |    StdErr |     StdDev |      Median |        Op/s | Scaled | Scaled-StdDev | Allocated |
----------------- |------------ |---------- |----------- |------------ |------------ |------- |-------------- |---------- |
 StringNoSegments |  44.5237 ns | 0.0542 ns |  0.2971 ns |  44.5933 ns | 22459939.85 |   1.00 |          0.00 |       0 B |
   StringSegments | 176.9724 ns | 2.6630 ns | 14.5858 ns | 170.0220 ns |  5650597.25 |   3.97 |          0.32 |      56 B |
   SpanNoSegments |  32.5865 ns | 0.0767 ns |  0.4203 ns |  32.4675 ns | 30687545.28 |   0.73 |          0.01 |       0 B |
     SpanSegments |  64.2064 ns | 0.0970 ns |  0.5310 ns |  64.0385 ns | 15574779.23 |   1.44 |          0.02 |       0 B |
```

Also, initial code had a redundant OR condition in the copy loop that yielded about half the perf gain here.

Will now try pre-computing `short`s and `int`s with dot segments.

cesarbs commented:
Pre-computing shorts and ints for dot segments didn't make a difference.

Waiting on #1489 to make changes in `Frame`. Please review that one.

davidfowl commented:
http://benchmarkdotnet.org/RulesOfBenchmarking.htm#avoid-dead-code-elimination

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/PathNormalizer.cs:
```diff
@@ -1,12 +1,195 @@
+            }
+        }
+
         public static string RemoveDotSegments(string path)
```
Will this method be removed?

cesarbs replied:
Yeah.

cesarbs commented:
I've changed the benchmarks a bit:

- Returning values instead of using member variables to prevent the JIT from optimizing the code away
- Removed the inner loops and changed the test path to a larger one
- Separate benchmarks for paths with `.` and `..` segments

Scaled values compared to the baseline changed, but the new implementation is still better than the current one:

```
                  Method |        Mean |    StdDev |      Median |       Op/s | Scaled | Scaled-StdDev | Allocated |
------------------------ |------------ |---------- |------------ |----------- |------- |-------------- |---------- |
        StringNoSegments | 150.3262 ns | 0.9072 ns | 150.4293 ns | 6652198.77 |   1.00 |          0.00 |       0 B |
 StringSingleDotSegments | 289.3416 ns | 3.9592 ns | 288.5401 ns | 3456122.28 |   1.92 |          0.03 |     152 B |
 StringDoubleDotSegments | 264.9764 ns | 2.4553 ns | 264.1891 ns | 3773921.21 |   1.76 |          0.02 |      40 B |
          SpanNoSegments | 137.2496 ns | 0.8896 ns | 136.7677 ns | 7285998.28 |   0.91 |          0.01 |       0 B |
   SpanSingleDotSegments | 178.6626 ns | 0.9520 ns | 178.2067 ns | 5597141.61 |   1.19 |          0.01 |       0 B |
   SpanDoubleDotSegments | 218.8283 ns | 1.2004 ns | 218.3675 ns |  4569792.9 |   1.46 |          0.01 |       0 B |
```

arespr commented:
I tried to add optimization before switch instruction. I have strange tests results - StringNoSegments is the fastest. Processor 4702mq
```
const byte mask = '.' & '/'; 
if ((ch1 & mask) == mask) // it is valid for only 16 characters: '.', '/', '>', '?', 'n', 'o', '~', '', '®', '¯', '¾', '¿', 'î', 'ï', 'þ', 'ÿ', 

{
    byte ch2, ch3, ch4;

    switch (end - src)
```

***Pr Before***
```
 |                  Method |        Mean |    StdErr |     StdDev |        Op/s | Scaled | Scaled-StdDev | Allocated |
 |------------------------ |------------ |---------- |----------- |------------ |------- |-------------- |---------- |
 |        StringNoSegments |  78.7066 ns | 0.5709 ns |  3.1268 ns | 12705421.54 |   1.00 |          0.00 |       0 B |
 | StringSingleDotSegments | 224.9039 ns | 0.6595 ns |  3.6122 ns |  4446342.96 |   2.86 |          0.12 |     152 B |
 | StringDoubleDotSegments | 213.5637 ns | 2.0682 ns | 11.3280 ns |  4682442.99 |   2.72 |          0.18 |      40 B |
 |          SpanNoSegments | 108.3500 ns | 0.6101 ns |  3.3418 ns |  9229346.75 |   1.38 |          0.07 |       0 B |
 |   SpanSingleDotSegments | 138.3004 ns | 0.8779 ns |  4.8082 ns |  7230634.62 |   1.76 |          0.09 |       0 B |
 |   SpanDoubleDotSegments | 189.7800 ns | 2.1132 ns | 11.5743 ns |  5269258.86 |   2.41 |          0.17 |       0 B |
```

***After***
```
 |                  Method |        Mean |    StdDev |        Op/s | Scaled | Scaled-StdDev | Allocated |
 |------------------------ |------------ |---------- |------------ |------- |-------------- |---------- |
 |        StringNoSegments |  78.8242 ns | 3.2197 ns | 12686457.68 |   1.00 |          0.00 |       0 B |
 | StringSingleDotSegments | 224.9548 ns | 3.6692 ns |  4445338.28 |   2.86 |          0.13 |     152 B |
 | StringDoubleDotSegments | 195.8797 ns | 1.5652 ns |   5105173.6 |   2.49 |          0.10 |      40 B |
 |          SpanNoSegments | 108.0713 ns | 0.9436 ns |   9253150.4 |   1.37 |          0.06 |       0 B |
 |   SpanSingleDotSegments | 139.9129 ns | 1.0133 ns |  7147304.26 |   1.78 |          0.07 |       0 B |
 |   SpanDoubleDotSegments | 184.2883 ns | 0.9511 ns |  5426279.89 |   2.34 |          0.10 |       0 B |
```

cesarbs commented:
This https://github.com/aspnet/KestrelHttpServer/blob/75bd6b55cf0e3c97022e0e6cf6abf68e5daa0365/src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/PathNormalizer.cs#L176 was wrong because it could read one byte past the end of the buffer. I fixed that and perf regressed (scaled `SpanNoSegments` when up to 1.14):

```
                  Method |        Mean |    StdErr |     StdDev |      Median |        Op/s | Scaled | Scaled-StdDev | Allocated |
------------------------ |------------ |---------- |----------- |------------ |------------ |------- |-------------- |---------- |
        StringNoSegments |  69.4170 ns | 0.7356 ns |  4.0293 ns |  70.2032 ns | 14405697.08 |   1.00 |          0.00 |       0 B |
 StringSingleDotSegments | 215.4227 ns | 1.8399 ns | 10.0774 ns | 212.9054 ns |  4642036.59 |   3.11 |          0.23 |     152 B |
 StringDoubleDotSegments | 167.1577 ns | 0.7056 ns |  3.8646 ns | 165.0735 ns |  5982373.05 |   2.42 |          0.15 |      40 B |
          SpanNoSegments |  79.0088 ns | 0.4797 ns |  2.6274 ns |  78.6207 ns | 12656821.82 |   1.14 |          0.08 |       0 B |
   SpanSingleDotSegments |  99.9484 ns | 0.6768 ns |  3.7070 ns |  98.1957 ns | 10005161.08 |   1.44 |          0.10 |       0 B |
   SpanDoubleDotSegments | 130.8632 ns | 0.5234 ns |  2.8670 ns | 129.7166 ns |  7641568.16 |   1.89 |          0.12 |       0 B |
```

However, since the `Span`-based code is only being used on origin-form targets, I was able to remove a bunch of code and brought most of the perf advantage back.

Origin-form means the path always starts with a `'/'`. Also, the way the code is written ensures `src` is always pointing to a `'/'` at the start of the main loop.

Initially I removed all conditions that checked for a starting `'.'`:

```
                  Method |        Mean |    StdDev |      Median |        Op/s | Scaled | Scaled-StdDev | Allocated |
------------------------ |------------ |---------- |------------ |------------ |------- |-------------- |---------- |
        StringNoSegments |  70.7242 ns | 1.0718 ns |  70.4569 ns | 14139438.15 |   1.00 |          0.00 |       0 B |
 StringSingleDotSegments | 215.3916 ns | 6.1774 ns | 212.3913 ns |  4642707.08 |   3.05 |          0.10 |     152 B |
 StringDoubleDotSegments | 165.8617 ns | 3.6387 ns | 164.1380 ns |  6029119.14 |   2.35 |          0.06 |      40 B |
          SpanNoSegments |  75.9759 ns | 1.4254 ns |  75.8074 ns | 13162072.31 |   1.07 |          0.03 |       0 B |
   SpanSingleDotSegments |  95.9031 ns | 2.9483 ns |  95.1903 ns | 10427190.53 |   1.36 |          0.05 |       0 B |
   SpanDoubleDotSegments | 123.2213 ns | 2.3020 ns | 122.2172 ns |  8115482.97 |   1.74 |          0.04 |       0 B |
```

Then I removed all the `ch1 == ByteSlash` checks, which were redundant:

```
                  Method |        Mean |    StdDev |      Median |        Op/s | Scaled | Scaled-StdDev | Allocated |
------------------------ |------------ |---------- |------------ |------------ |------- |-------------- |---------- |
        StringNoSegments |  71.5367 ns | 2.1474 ns |  71.8455 ns | 13978844.06 |   1.00 |          0.00 |       0 B |
 StringSingleDotSegments | 215.8189 ns | 6.5263 ns | 213.4131 ns |  4633513.68 |   3.02 |          0.13 |     152 B |
 StringDoubleDotSegments | 167.6640 ns | 3.8528 ns | 166.2176 ns |  5964309.33 |   2.35 |          0.09 |      40 B |
          SpanNoSegments |  67.2115 ns | 0.7956 ns |  66.8048 ns | 14878407.91 |   0.94 |          0.03 |       0 B |
   SpanSingleDotSegments |  84.0777 ns | 1.2682 ns |  83.5769 ns | 11893763.61 |   1.18 |          0.04 |       0 B |
   SpanDoubleDotSegments | 114.0320 ns | 2.6034 ns | 112.7485 ns |  8769469.66 |   1.60 |          0.06 |       0 B |
```

Perf is unaffected in other benchmarks, and allocations remain the same:

## RequestParsing

### dev

```
                        Method |          Mean |      StdDev |       Op/s | Scaled | Scaled-StdDev | Allocated |
------------------------------ |-------------- |------------ |----------- |------- |-------------- |---------- |
          PlaintextTechEmpower | 1,172.5749 ns |  24.9941 ns |  852824.01 |   1.00 |          0.00 |     349 B |
          PlaintextAbsoluteUri | 2,077.4782 ns |  65.5862 ns |  481352.82 |   1.77 |          0.07 |     692 B |
 PipelinedPlaintextTechEmpower |   963.5593 ns |  28.0590 ns | 1037818.89 |   0.82 |          0.03 |     348 B |
                    LiveAspNet | 2,273.8135 ns |  60.8642 ns |  439789.81 |   1.94 |          0.06 |   1.07 kB |
           PipelinedLiveAspNet | 2,166.0772 ns |  68.0248 ns |  461664.06 |   1.85 |          0.07 |   1.07 kB |
                       Unicode | 3,667.5794 ns | 135.2705 ns |   272659.4 |   3.13 |          0.13 |   1.85 kB |
              UnicodePipelined | 3,544.3220 ns |  98.1693 ns |  282141.41 |   3.02 |          0.10 |   1.86 kB |
```

### cesarbs/1448

```
                        Method |          Mean |     StdErr |      StdDev |       Op/s | Scaled | Scaled-StdDev | Allocated |
------------------------------ |-------------- |----------- |------------ |----------- |------- |-------------- |---------- |
          PlaintextTechEmpower | 1,153.2268 ns |  5.0885 ns |  27.8708 ns |  867132.13 |   1.00 |          0.00 |     349 B |
          PlaintextAbsoluteUri | 2,010.9973 ns | 23.8681 ns | 130.7308 ns |   497265.7 |   1.74 |          0.12 |     692 B |
 PipelinedPlaintextTechEmpower |   961.2483 ns |  3.7872 ns |  20.7431 ns | 1040313.96 |   0.83 |          0.03 |     348 B |
                    LiveAspNet | 2,294.4870 ns | 10.6202 ns |  58.1692 ns |  435827.27 |   1.99 |          0.07 |   1.07 kB |
           PipelinedLiveAspNet | 2,090.3801 ns | 16.6631 ns |  91.2678 ns |   478381.9 |   1.81 |          0.09 |   1.07 kB |
                       Unicode | 3,652.8222 ns | 21.4497 ns | 117.4851 ns |  273760.93 |   3.17 |          0.12 |   1.85 kB |
              UnicodePipelined | 3,515.5588 ns | 14.8895 ns |  81.5534 ns |   284449.8 |   3.05 |          0.10 |   1.86 kB |
```

## FrameParsingOverhead

### dev

```
                      Method |        Mean |     StdDev |       Op/s | Scaled | Scaled-StdDev | Allocated |
---------------------------- |------------ |----------- |----------- |------- |-------------- |---------- |
          FrameOverheadTotal | 439.6411 ns | 13.7209 ns | 2274582.75 |   1.00 |          0.00 |     359 B |
    FrameOverheadRequestLine | 190.7454 ns |  6.7451 ns | 5242589.43 |   0.43 |          0.02 |      48 B |
 FrameOverheadRequestHeaders | 296.3084 ns | 10.8587 ns | 3374862.25 |   0.67 |          0.03 |     311 B |
```

### cesarbs/1448

```
                      Method |        Mean |    StdDev |      Median |       Op/s | Scaled | Scaled-StdDev | Allocated |
---------------------------- |------------ |---------- |------------ |----------- |------- |-------------- |---------- |
          FrameOverheadTotal | 435.7483 ns | 8.0143 ns | 432.7154 ns | 2294903.04 |   1.00 |          0.00 |     359 B |
    FrameOverheadRequestLine | 186.9259 ns | 8.6856 ns | 184.8890 ns | 5349712.06 |   0.43 |          0.02 |      48 B |
 FrameOverheadRequestHeaders | 297.3806 ns | 9.8844 ns | 292.7018 ns | 3362694.07 |   0.68 |          0.03 |     311 B |
```

cesarbs commented:
I found some issues in the original `string`-based code, and decided to re-write it following the same algorithm used for `Span`s (with more checks because it can't make assumptions about `'/'`). It turns out the new string code is faster than the `Span` code for the no dot segments case 😄 But the `Span` code is now as fast as the original code for that case, and faster when there are segments plus it never allocates.

Here are the benchmark results:

```
                     Method |        Mean |    StdErr |     StdDev |        Op/s | Scaled | Scaled-StdDev | Allocated |
--------------------------- |------------ |---------- |----------- |------------ |------- |-------------- |---------- |
        StringNoSegmentsOld |  65.0562 ns | 0.3028 ns |  1.6583 ns | 15371321.34 |   1.00 |          0.00 |       0 B |
 StringSingleDotSegmentsOld | 220.6973 ns | 1.0482 ns |  5.7413 ns |  4531092.11 |   3.39 |          0.12 |     152 B |
 StringDoubleDotSegmentsOld | 174.7711 ns | 0.8494 ns |  4.6525 ns |  5721770.52 |   2.69 |          0.10 |      40 B |
           StringNoSegments |  59.1011 ns | 0.1652 ns |  0.9048 ns | 16920167.24 |   0.91 |          0.03 |       0 B |
    StringSingleDotSegments | 168.1116 ns | 1.1702 ns |  6.4095 ns |  5948428.68 |   2.59 |          0.12 |     343 B |
    StringDoubleDotSegments | 195.3203 ns | 2.0398 ns | 11.1725 ns |   5119795.1 |   3.00 |          0.18 |     255 B |
             SpanNoSegments |  64.9553 ns | 0.2893 ns |  1.5843 ns | 15395199.35 |   1.00 |          0.03 |       0 B |
      SpanSingleDotSegments |  82.1222 ns | 0.1902 ns |  1.0418 ns | 12176969.34 |   1.26 |          0.03 |       0 B |
      SpanDoubleDotSegments | 135.2150 ns | 0.3467 ns |  1.8988 ns |  7395628.19 |   2.08 |          0.06 |       0 B |
```

cesarbs commented:
Most of my benchmarks so far were wrong.

Each `[Benchmark]` method is executed multiple times per iteration. The ones that were mutating the input were doing their work only once.

![image](https://cloud.githubusercontent.com/assets/924164/24027087/4361b71a-0a82-11e7-8f44-08424a8e4848.png)
(https://commons.wikimedia.org/wiki/File:Paris_Tuileries_Garden_Facepalm_statue.jpg)

So I fixed that.

I also changed the string version to bastardize `string` and mutate it. It's fast. It can actually take the same shortcuts as the `Span` version since it's only used on absolute form targets and we get the path from the `Uri` that builds. I had thought I would need more checks in there because of authority-form, but we don't remove dot segments from that.

Here are the new numbers (had to artificially copy buffers in the benchmarks for the old code so it did the same amount of work as the new stuff):

```
                     Method |        Mean |    StdErr |     StdDev |      Median |       Op/s | Scaled | Scaled-StdDev | Allocated |
--------------------------- |------------ |---------- |----------- |------------ |----------- |------- |-------------- |---------- |
        StringNoSegmentsOld | 188.1918 ns | 1.1323 ns |  6.2018 ns | 187.7370 ns | 5313728.74 |   1.00 |          0.00 |       0 B |
 StringSingleDotSegmentsOld | 417.5682 ns | 5.0027 ns | 27.4011 ns | 408.5006 ns | 2394818.67 |   2.22 |          0.16 |     152 B |
 StringDoubleDotSegmentsOld | 361.3352 ns | 1.0959 ns |  6.0027 ns | 359.5030 ns | 2767513.67 |   1.92 |          0.07 |      40 B |
           StringNoSegments | 176.3827 ns | 0.7492 ns |  4.1038 ns | 174.3564 ns | 5669491.07 |   0.94 |          0.04 |       0 B |
    StringSingleDotSegments | 216.0178 ns | 0.7891 ns |  4.3223 ns | 214.8109 ns | 4629248.85 |   1.15 |          0.04 |       0 B |
    StringDoubleDotSegments | 303.6101 ns | 0.8111 ns |  4.4425 ns | 302.8516 ns | 3293698.43 |   1.61 |          0.06 |       0 B |
             SpanNoSegments | 160.8814 ns | 0.3020 ns |  1.6539 ns | 160.7331 ns | 6215758.88 |   0.86 |          0.03 |       0 B |
      SpanSingleDotSegments | 241.6093 ns | 0.3045 ns |  1.6680 ns | 241.7209 ns | 4138912.92 |   1.29 |          0.04 |       0 B |
      SpanDoubleDotSegments | 276.3474 ns | 1.6365 ns |  8.9635 ns | 274.4087 ns | 3618634.21 |   1.47 |          0.07 |       0 B |
```

I'm not going to keep the old code around though, so here are the benchmarks just for the new stuff:

```
                  Method |        Mean |    StdErr |     StdDev |       Op/s | Scaled | Scaled-StdDev | Allocated |
------------------------ |------------ |---------- |----------- |----------- |------- |-------------- |---------- |
        StringNoSegments | 175.8846 ns | 0.5428 ns |  2.9732 ns | 5685546.75 |   1.00 |          0.00 |       0 B |
 StringSingleDotSegments | 216.2468 ns | 0.4660 ns |  2.5522 ns |  4624345.1 |   1.23 |          0.02 |       0 B |
 StringDoubleDotSegments | 304.5847 ns | 0.9166 ns |  5.0202 ns | 3283158.51 |   1.73 |          0.04 |       0 B |
          SpanNoSegments | 164.2495 ns | 3.1072 ns | 17.0188 ns | 6088298.27 |   0.93 |          0.10 |       0 B |
   SpanSingleDotSegments | 246.5714 ns | 3.9175 ns | 21.4569 ns | 4055620.52 |   1.40 |          0.12 |       0 B |
   SpanDoubleDotSegments | 275.1196 ns | 1.4397 ns |  7.8854 ns | 3634782.79 |   1.56 |          0.05 |       0 B |
```

cesarbs commented:
We don't need the string-based code at all. It was only used on absolute-form targets, but we use `Uri.TryParse` on those and that removes dot segments.

Keeping the benchmarks for `Span` to measure future improvements.
